Task 2.11

DmKoshelek opened the pull request.

Commit:
Task 2.3
Implement integration test for JdbcPhoneDao

Commit:
Task 2.3
Code refactoring

Commit:
Task 2.5
Updated phoneList.jsp
Added to phoneDao ability to take sorted list

Commit:
Task 2.5
Add pagination
Add ordering
Add searh

Commit:
Task 2.5
Realized HttpSeesionCartService
Add tests for HttpSeesionCartService

Commit:
Task 2.5
Realized AjaxCartController
Add tests for AjaxCartController
Fix HttpSessionCartService

Commit:
Task 2.5
Fixed JdbcPhoneDao for get phones without colors
Fixed AjaxCartController for non-numerical values
Added function "Add to cart"

OEAndreiIvanov commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,95 @@
-    public void addPhone(Long phoneId, Long quantity) {
-        throw new UnsupportedOperationException("TODO");
+    public BigDecimal getCartCost() {
+        BigDecimal cartCost = (BigDecimal) httpSession.getAttribute(PHONES_COST_IN_CART_ATTRIBUTE_NAME);
```
Зачем хранить стоимость как отдельный атрибут в сессии?

OEAndreiIvanov commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,95 @@
+        cart.addPhone(phoneId, quantity);
+        BigDecimal cartCost = getCartCost();
+        BigDecimal phonePrice = addedPhone.getPrice();
+        if (phonePrice == null) {
```
В каких случаях такое может произойти? Почему PhoneNotFoundException?

OEAndreiIvanov commented on core/src/main/java/com/es/core/cart/HttpSessionCartService.java:
```diff
@@ -1,28 +1,95 @@
+        List<Phone> settedPhones = phoneDao.getPhonesByIdList(new ArrayList<>(items.keySet()));
+        cart.setItems(items);
+
+        BigDecimal cartCost = settedPhones.stream()
```
Можно написать один метод для пересчета всей карточки

OEAndreiIvanov commented on core/src/main/java/com/es/core/model/phone/JdbcPhoneDao.java:
```diff
@@ -1,27 +1,253 @@
 
     public Optional<Phone> get(final Long key) {
-        throw new UnsupportedOperationException("TODO");
+        List<Phone> phones = jdbcTemplate.query(SELECT_PHONE_QUERY, new PhoneListResultSetExtractor(), key);
```
extractor можно вынести в синглтон

OEAndreiIvanov commented on core/src/main/java/com/es/core/model/phone/JdbcPhoneDao.java:
```diff
@@ -1,27 +1,253 @@
+
+    @Override
+    public List<Phone> getPhonesByIdList(List<Long> idList) {
+        String delimetedList = idList.stream().map(Object::toString).collect(Collectors.joining(", ", "(", ")"));
```
NamedParameterJdbcTemplate не поддерживает такой запрос через один бинд параметр?

OEAndreiIvanov commented on web/src/main/test/com/es/phoneshop/web/proxy/CartServiceProxy.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Map;
+
+
+public class CartServiceProxy implements CartService {
```
Для мокито тоже будет прокси?

OEAndreiIvanov commented on web/src/main/test/resources/context/testContext-web.xml:
```diff
@@ -0,0 +1,11 @@
+       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd http://www.springframework.org/schema/context http://www.springframework.org/schema/context/spring-context.xsd http://www.springframework.org/schema/mvc http://www.springframework.org/schema/mvc/spring-mvc.xsd">
+
+    <mvc:annotation-driven/>
+    <bean id="ajaxCartController" class="com.es.phoneshop.web.controller.AjaxCartController"/>
```
Излишне объявлять так каждый контроллер

Commit:
Task 2.6
Added product details page
Rewrited tests for CartService using Mockito
Some code refactoring

Commit:
Task 2.7
Added cart page
Added remove product function

Commit:
Code Refactoring

Commit:
Code Refactoring

Commit:
Code refacroring
Rewrited tests using Mockito
Fixed test classpath

Commit:
Task 2.7

DmKoshelek committed:
Code refactoring

DmKoshelek committed:
Task 2.7
Code refactoring
Bug fixed for updating empty cart

DmKoshelek committed:
Task 2.7
Code refactoring
Fixed bugs

DmKoshelek committed:
Task 2.8
Added JdbcOrderDao
Added tests for JdbcOrderDao

DmKoshelek committed:
Task 2.8
Added order page
Added validation of stock

DmKoshelek committed:
Task 2.9
Added order overview page

OEAndreiIvanov commented on web/src/main/java/com/es/phoneshop/web/controller/pages/OrderOverviewPageController.java:
```diff
@@ -1,9 +1,38 @@
+    @GetMapping
+    public String getOrderOverview(@PathVariable("id") Long id, Model model) throws PageNotFoundException{
+        Optional<Order> orderOptional = orderService.getOrder(id);
+        if(!orderOptional.isPresent()){
```
orElseThrow, code format

OEAndreiIvanov commented on core/src/test/java/com/es/core/cart/HttpSessionCartServiceTest.java:
```diff
@@ -257,6 +257,13 @@ public void deleteOutOfStockCheck() {
+    @Test
+    public void clearCart() {
+        cartService.clearCart();
+        verify(mockCart, atLeastOnce()).setItems(anyList());
```
Проверь, что после вызова метода clear, карточка стала пустой и с нулевым price, вместо того что вызвались те методы, которые у тебя в коде

OEAndreiIvanov commented on core/src/main/java/com/es/core/dao/phone/JdbcPhoneDao.java:
```diff
@@ -187,4 +194,10 @@ public int phoneCountByQuery(String query) {
 
+    @Override
+    public List<Stock> getStocks(List<Long> phoneIdList) {
+        Map namedParameters = Collections.singletonMap("phoneIdList", phoneIdList);
```
raw type

OEAndreiIvanov commented on core/src/main/java/com/es/core/model/order/OrderItem.java:
```diff
@@ -40,6 +56,10 @@ public void setQuantity(final Long quantity) {
         this.quantity = quantity;
     }
 
+    public BigDecimal getTotal() {
```
Если уж логика для reprice где-то в сервисе, пусть он же и считает такие вещи

OEAndreiIvanov commented on web/src/main/java/com/es/phoneshop/web/controller/pages/OrderPageController.java:
```diff
@@ -1,26 +1,70 @@
+        try {
+            orderService.placeOrder(order);
+            return "redirect:/orderOverview/" + Long.toString(order.getId());
+        } catch (OutOfStockException e) {
```
Хотелось бы такую проверку в валидатор и если что то добавить в global errors

DmKoshelek committed:
Task 2.10
Added admin page

DmKoshelek committed:
Task 2.10
Fixed a bug with cyrilic symbols in the order form
Code refactoring

DmKoshelek committed:
Task 2.11
Add login for admin

DmKoshelek committed:
Task 2.11
Fixed logout

DmKoshelek committed:
Task 2.11
Fixed AjaxCartControllerTest
